tests/codec.rs

```diff
@@ -0,0 +1,51 @@
+use std::{collections::BTreeMap, iter};
+
+use ipld_core::{
+    cid::Cid,
+    codec::{Codec, Links},
+    ipld,
+    ipld::Ipld,
+};
+use ipld_dagpb::{DagPbCodec, PbNode};
+
+fn gen_ipld_node() -> Ipld {
+    let cid = Cid::try_from("bafkreie74tgmnxqwojhtumgh5dzfj46gi4mynlfr7dmm7duwzyvnpw7h7m").unwrap();
+    let pb_link = ipld!({
+        "Hash": cid,
+        "Name": "block",
+        "Tsize": 13,
+    });
+
+    let links: Vec<Ipld> = vec![pb_link.clone(), pb_link];
+    let mut pb_node = BTreeMap::<String, Ipld>::new();
+    pb_node.insert("Data".to_string(), b"Here is some data\n".to_vec().into());
+    pb_node.insert("Links".to_string(), links.into());
+    pb_node.into()
+}
+
+#[test]
+fn test_codec_ipld_encode_decode() {
+    let ipld = gen_ipld_node();
+    let bytes = DagPbCodec::encode_to_vec(&ipld).unwrap();
+    let ipld2 = DagPbCodec::decode_from_slice(&bytes).unwrap();
+    assert_eq!(ipld, ipld2);
+}
+
+#[test]
+fn test_codec_pbnode_encode_decode() {
+    let ipld = gen_ipld_node();
+    let bytes = DagPbCodec::encode_to_vec(&ipld).unwrap();
+    let pb_node: PbNode = DagPbCodec::decode_from_slice(&bytes).unwrap();
```

Comment:
So how is the inference working here, does it choose `Ipld` by default (as in the `let ipld2 = DagPbCodec::decode_from_slice above` since it comes first in the implementation? but it'll do a `PbNode` for you optionally if you type it and let it make a clear inference? What's the mechanism that sets up the default when there's no type to infer, or is this done in the trait definition itself?

Reply:
Types are always inferred. If it cannot infer a type, it will give you a compiler error.

Above the `let ipld2` is inferred as the `assert!()` expects the types to be the same, and the `ipld` is of type `Ipld`.